pkgs/servers/web-apps/bookstack/default.nix

```diff
@@ -0,0 +1,38 @@
+{ pkgs, system, lib, fetchFromGitHub, dataDir ? "/var/lib/bookstack" }:
+
+let
+  package = (import ./composition.nix {
+    inherit pkgs system;
+    noDev = true; # Disable development dependencies
+  }).overrideAttrs (attrs : {
+    installPhase = attrs.installPhase + ''
+      rm -R $out/storage $out/public/uploads
```

Comment:
Hmm... what about the `bootstrap/` folder?

Reply:
I had some issues with that, I can't quite remember but it doesn't like the bootstrap/cache folder being symlinked.

Reply:
[This issue on laravel/framework](https://github.com/laravel/framework/issues/12526) and [this one at invoiceninja](https://github.com/invoiceninja/invoiceninja/issues/1830) are about laravel having issues with symlinks...

Reply:
The `bootstrap/cache` folder is mutable and in the immutable nix store. My assumption is that you are having some silent failures, perhaps.

Reply:
Some more info from [one of the issues I linked above](https://github.com/invoiceninja/invoiceninja/issues/1830): 

> It's sad that Lavarel does not seem to work with symlinks. At least, as soon as I start to symlink /bootstrap/cache or /bootstrap everything starts to fall apart. Fortunately this is not the case for /storage, so I was successfully able to move this in /app/data/storage, to let the backup take care of it.

That's why I can't use the same trick I've used for the `storage` dir for `bootstrap/cache`. It fails when you symlink it. Maybe you have a better idea here? 

I guess there might be a slight performance penalty, but I've not noticed any issues.

Reply:
Some more info from [one of the issues I linked above](https://github.com/invoiceninja/invoiceninja/issues/1830): 

> It's sad that Lavarel does not seem to work with symlinks. At least, as soon as I start to symlink /bootstrap/cache or /bootstrap everything starts to fall apart. Fortunately this is not the case for /storage, so I was successfully able to move this in /app/data/storage, to let the backup take care of it.

That's why I can't use the same trick I've used for the `storage` dir for `bootstrap/cache`. It fails when you symlink it. Maybe you have a better idea here? 

I guess there might be a slight performance penalty, but I've not noticed any issues.

Reply:
I believe the following environment variables are what you need for `bootstrap/cache`: `APP_SERVICES_CACHE`, `APP_PACKAGES_CACHE`, `APP_CONFIG_CACHE`, `APP_ROUTES_CACHE`, and `APP_EVENTS_CACHE`.

You could set them as follows:
```
APP_SERVICES_CACHE=/var/cache/bookstack/services.php
APP_PACKAGES_CACHE=/var/cache/bookstack/packages.php
APP_CONFIG_CACHE=/var/cache/bookstack/config.php
APP_ROUTES_CACHE=/var/cache/bookstack/routes-v7.php
APP_EVENTS_CACHE=/var/cache/bookstack/events.php
```

This would allow you to run initialization commands like `php artisan config:cache`, `php artisan route:cache`, `php artisan view:cache`, etc... which should have a **significant** improvement. Additionally, this *may* solve the issue I mentioned about `php artisan` commands not working. Just make sure `/var/cache` has strict permissions (`0700`) and is owned exclusively for read/write access to the `bookstack` user.

And to be clear - no I have not tested any of this... **but** I do a fair bit of (imperative) devops for Laravel developers.

Reply:
Sounds great, thanks. I'll try to set those and come back to you later.
For now, I've converted this to a draft PR as long as we're working on this.

> And to be clear - no I have not tested any of this... but I do a fair bit of (imperative) devops for Laravel developers.

Great, I had absolutely 0 experience with laravel, so thanks for your help!

Reply:
Please have another look, @aanderse, I've managed to add a cache dir with 700 permissions and it seems to work for me (including `php artisan` commands)

Reply:
Great, will do. How do you run `artisan` commands as a sysadmin with this install?

Reply:
I guess one would have to set `users.users.bookstack.shell` to `pkgs.bash` or something and then cd into the bookstack dir in the nix-store and run them. I don't really see another way...

Reply:
We can take inspiration from prior work in NixOS to solve this problem. If we look through the [nextcloud module](https://github.com/NixOS/nixpkgs/blob/nixos-20.09/nixos/modules/services/web-apps/nextcloud.nix#L32) we can see a great solution that is easy to adapt into *something* like this:
```
let
  artisan = pkgs.writeScriptBin "bookshelf" ''
    #! ${pkgs.runtimeShell}
    cd ${cfg.package}
    sudo=exec
    if [[ "$USER" != bookshelf ]]; then
      sudo='exec /run/wrappers/bin/sudo -u bookshelf'
    fi
    $sudo ${pkgs.php}/bin/php artisan $*
  '';
in
  environment.systemPackages = [ artisan ];
```

From here it looks like the sysadmin simply runs `bookshelf cache:clear` and good to go :+1: .

Reply:
Works for me now! 
```
❯ bookstack cache:clear
Application cache cleared!
```

Reply:
That's great. I'm really happy with how this module has shaped up. Thanks for all the work on it.

In ended up doing some testing with `Laravel` and can confirm the ideas in this PR seem to work really well. 👍